File: network/rpc-examples/minirpc2/minirpc/rpc_channel.cc
// Copyright (c) 2017 ASMlover. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
//
//  * Redistributions of source code must retain the above copyright
//    notice, this list ofconditions and the following disclaimer.
//
//  * Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in
//    the documentation and/or other materialsprovided with the
//    distribution.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.
#include <iostream>
#include "rpc.pb.h"
#include "rpc_channel.h"

namespace minirpc {

RpcChannel::RpcChannel(asio::io_service& io_service, asio::tcp::socket&& socket)
  : io_service_(io_service)
  , socket_(std::move(socket)) {
}

RpcChannel::~RpcChannel(void) {
}

void RpcChannel::CallMethod(const gpb::MethodDescriptor* method, gpb::RpcController* /*controller*/,
    const gpb::Message* request, gpb::Message* response, gpb::Closure* done) {
  std::int64_t id = ++id_;

  minirpc::RpcMessage message;
  message.set_type(TYPE_REQUEST);
  message.set_id(id);
  message.set_service(method->service()->name());
  message.set_method(method->name());
  message.set_request(request->SerializeAsString());

  {
    OutstandingCall out{response, done};
    std::unique_lock<std::mutex> guard(mutex_);
    outstandings_.insert(std::make_pair(id, out));
  }

  do_write(message);
}

void RpcChannel::start(void) {
  do_read_header();
}

void RpcChannel::close(void) {
  auto self(shared_from_this());
  io_service_.post([this, self] { socket_.close(); });
}

void RpcChannel::do_read_header(void) {
  buffer_.resize(4);
  auto self(shared_from_this());
  asio::async_read(socket_, asio::buffer(buffer_),
      [this, self](const boost::system::error_code& ec, std::size_t n) {
        if (!ec && n == 4) {
          std::size_t len = 0;
          std::memcpy(&len, buffer_.data(), 4);
          do_read_body(len);
        }
        else {
          socket_.close();
        }
      });
}

void RpcChannel::do_read_body(std::size_t len) {
  buffer_.resize(len);
  auto self(shared_from_this());
  asio::async_read(socket_, asio::buffer(buffer_),
      [this, self, len](const boost::system::error_code& ec, std::size_t n) {
        if (!ec && len == n) {
          minirpc::RpcMessage message;
          if (decode(buffer_, message))
            handle_message(message);

          do_read_header();
        }
        else {
          socket_.close();
        }
      });
}

void RpcChannel::do_write(const RpcMessage& message) {
  std::vector<char> buf;
  encode(message, buf);

  auto self(shared_from_this());
  asio::async_write(socket_, asio::buffer(buf),
      [this, self](const boost::system::error_code& ec, std::size_t /*n*/) {
        if (ec) {
          std::cerr << "RpcChannel::do_write - write RpcMessage failed ..." << std::endl;
          socket_.close();
        }
      });
}

void RpcChannel::handle_message(const RpcMessage& message) {
  if (message.type() == TYPE_REQUEST) {
    if (service_) {
      const auto* desc = service_->GetDescriptor();
      const auto* method = desc->FindMethodByName(message.method());
      if (method) {
        auto* request = service_->GetRequestPrototype(method).New();
        request->ParseFromString(message.request());
        service_->CallMethod(method, nullptr, request, nullptr, nullptr);
        // TODO: solve response

        delete request;
      }
    }
  }
  else if (message.type() == TYPE_RESPONSE) {
    std::int64_t id = message.id();
    OutstandingCall out{};

    {
      std::unique_lock<std::mutex> guard(mutex_);
      auto it = outstandings_.find(id);
      if (it != outstandings_.end()) {
        out = it->second;
        outstandings_.erase(it);
      }
    }

    if (out.response) {
      out.response->ParseFromString(message.response());
      if (out.done)
        out.done->Run();
      delete out.response;
    }
  }
}

void RpcChannel::done_callback(gpb::Message* response, std::int64_t id) {
  minirpc::RpcMessage message;
  message.set_type(TYPE_RESPONSE);
  message.set_id(id);
  message.set_response(response->SerializeAsString());
  do_write(message);

  delete response;
}

void RpcChannel::encode(const RpcMessage& msg, std::vector<char>& buf) {
  const int bytes = msg.ByteSize();
  const int len = bytes + 4; // RPC0
  const int ntotal = len + 4;

  buf.resize(ntotal);
  std::memcpy(&buf[0], &len, sizeof(len));
  std::memcpy(&buf[4], "RPC0", 4);
  msg.SerializeWithCachedSizesToArray((std::uint8_t*)&buf[8]);
}

bool RpcChannel::decode(const std::vector<char>& buf, RpcMessage& msg) {
  if (std::memcmp(buf.data(), "RPC0", 4) == 0) {
    if (msg.ParseFromArray(&buf[4], buf.size() - 4))
      return true;
  }

  return false;
}

}

File: server/NyxEngine/Core/Rpc/RpcChannel.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
#
# Copyright (c) 2017 ASMlover. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list ofconditions and the following disclaimer.
#
#  * Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materialsprovided with the
#    distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

from struct import pack, unpack
from google.protobuf import service
from Log.LogManager import LogManager
from Rpc import Request

class RpcController(service.RpcController):
    def __init__(self, channel):
        super(RpcController, self).__init__()
        self.rpc_channel = channel

class RpcChannel(service.RpcChannel):
    def __init__(self, rpc_service, conn):
        super(RpcChannel, self).__init__()
        self.logger = LogManager.get_logger('Rpc.RpcChannel')

        self.rpc_service = rpc_service
        self.rpc_request = Request.Request()
        self.rpc_request_parser = Request.RequestParser()
        self.conn = conn
        self.conn.set_channel(self)
        self.controller = RpcController(self)
        self.conn_listeners = set()

        # user data information
        self.user_data = None
        self.encrypted = False
        self.compressed = False
        self.session_seed = None

    def set_max_databytes(self, data_bytes):
        self.rpc_request_parser.set_max_databytes(data_bytes)

    def reg_listener(self, listener):
        self.conn_listeners.add(listener)

    def unreg_listener(self, listener):
        self.conn_listeners.discard(listener)

    def get_peeraddr(self):
        if self.conn:
            return self.conn.get_peeraddr()
        return 'No connection attached'

    def sef_compressor(self, compressor):
        self.conn.set_compressor(compressor)

    def set_crypter(self, crypter):
        self.conn.set_crypter(crypter)

    def set_user_data(self, user_data):
        self.user_data = user_data

    def get_user_data(self):
        return self.user_data

    def set_session_seed(self, seed):
        self.session_seed = seed

    def get_session_seed(self):
        return self.session_seed

    def disconnect(self):
        """断开连接"""
        if self.conn:
            self.conn.disconnect()

    def on_disconnected(self):
        """连接(TcpSession)断开时回调"""
        for listener in list(self.conn_listeners):
            # 这里先拷贝到list是为了放置其他线程有使用unreg_*的情况
            if listener in self.conn_listeners:
                listener.on_channel_disconnected(self)
        self.rpc_request.reset()
        self.rpc_request_parser.reset()
        self.conn_listeners = None
        self.conn = None
        self.user_data = None

    def CallMethod(self, method, controller, request, response, done):
        """protobuf RpcChannel的重载，发送rpc"""
        index = method.index
        assert index < 65535
        data = request.SerializeToString()
        datalen = len(data) + 2
        self.conn.write_data(
                ''.join([pack('<I', datalen), pack('<H', index), data]))

    def on_request(self):
        # TODO:
        pass

    def on_input_data(self, data):
        # TODO:
        pass
